src/doc/rustc/src/jobserver.md

```diff
@@ -0,0 +1,55 @@
+# Jobserver
+
+Internally, `rustc` may take advantage of parallelism. `rustc` will coordinate
+with the build system calling it if a [GNU Make jobserver] is passed in the
+`MAKEFLAGS` environment variable.
+
+Starting with Rust 1.76.0, `rustc` will warn if a jobserver appears to be
+available but is not accessible, e.g.:
+
+```console
+$ echo 'fn main() {}' | MAKEFLAGS=--jobserver-auth=3,4 rustc -
+warning: failed to connect to jobserver from environment variable `MAKEFLAGS="--jobserver-auth=3,4"`: cannot open file descriptor 3 from the jobserver environment variable value: Bad file descriptor (os error 9)
+  |
+  = note: the build environment is likely misconfigured
+```
+
+## Integration with build systems
+
+The following subsections contain recommendations on how to integrate `rustc`
+with build systems so that the jobserver is handled appropriately.
+
+### GNU Make
+
+When calling `rustc` from GNU Make, it is recommended that all `rustc`
+invocations are marked as recursive in the `Makefile` (by prefixing the command
+line with the `+` indicator), so that GNU Make enables the jobserver for them.
+For instance:
+
+<!-- ignore-tidy-tab -->
+
+```make
+x:
+	+@echo 'fn main() {}' | rustc -
```

Comment:
`+` works in `make` itself, but not in some wrappers around it, like `cmake`.
The ["universal workaround"](https://github.com/rust-lang/rust/issues/46981#issuecomment-681090839) for such cases is to use `$(MAKE)` (literally) somewhere on the command line, then `make` will pass the jobserver even without `+`.

I haven't use the `$(shell ...)` construction mentioned below, but perhaps this method will work for it as well.

Reply:
Yeah, I use `$(MAKE)` in the example below (in the recursive Make case), but the GNU Make manual mentions `+` in the jobserver page as well as in the errors one (and I agree it is cleaner for this purpose than `$(MAKE)`), so I think it is best to keep `+` here. After all, we are in the section about GNU Make here, not other build systems.

But mentioning how to workaround it in CMake sounds like a good idea (even if it is just for their Makefile generator). After all, the intention of the document is to describe the recommendations for different build systems eventually, not just GNU Make.

I looked a bit into it, and I see issues like https://gitlab.kitware.com/cmake/cmake/-/issues/17781. However, I also found `JOB_SERVER_AWARE` from 3.28, documented in https://cmake.org/cmake/help/latest/command/ that actually adds a `+` in the generated `Makefile`s.

So I added that new CMake section in v4. Please take a look!